### PR TITLE
test(ut): added tests to check the existing behaviour for block/manager.go

### PR DESCRIPTION
**Pull request by arhamj**
## Overview

This MR has changes which potentially add the missing tests to `manager_test.go`. Currently, the tests only validate the existing behaviours, i.e.:

1. If there are A, B and C blocks which are in the pending blocks list and C has a marshal error. We currently throw an error.
2. If A is too big, A never gets submitted.
3. If A itself has a marshal error which is a unit case of (1), we also throw an error.

Closes: https://github.com/rollkit/rollkit/issues/1530

From the code, it looks like marshal errors would be caught earlier at `SaveBlock` to the store stage. That said I would like the following change:

- Offloading block aggregation from `da.go` to `manager.go` to avoid retries in cases where there are such issues and also for better error logging.

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

## Summary by CodeRabbit

- **Tests**
	- Enhanced testing for block submission, including cases with marshalling errors and size limit exceedances.
- **Chores**
	- Added a `mockery` command for generating mocks of the `Store` entity to facilitate testing.
	- Introduced a mock `Store` type to simulate store interactions in tests.

### CODE REVIEW

File: block/manager_test.go [hunk 1]
```diff
@@ -182,6 +182,43 @@ func TestSubmitBlocksToDA(t *testing.T) {
...
+		},
+		{
+			name: "A itself has a marshalling error, so A never gets submitted",
+			blocks: func() []*types.Block {
+				numBlocks, numTxs := 3, 5
+				blocks := make([]*types.Block, numBlocks)
+				for i := 0; i < numBlocks; i++ {
+					blocks[i] = types.GetRandomBlock(uint64(i+1), numTxs)
+				}
+				return blocks
```
Comment (MSevey) [inline block/manager_test.go:217]: You need to call `invalidateBlockHeader` on block A.

File: block/manager_test.go [hunk 2]
```diff
@@ -196,3 +233,14 @@ func TestSubmitBlocksToDA(t *testing.T) {
 		})
 	}
 }
+
+func InvalidateBlockHeader(block *types.Block) {
```
Comment (MSevey) [inline block/manager_test.go:237]: this should be unexported.

**Comment by arhamj** [comment]
> Thanks for picking up this issue! I left some comments to update the tests to match the desired state. There is a good change that once written properly, the tests will actually fail. This is one of the goals of the issue. If the tests are failing, we will need to investigate the changes required to the code itself to handle those edge cases and/ore discuss internally what the right direction to go is.

@MSevey can you please have a look again when you find time?

File: da/da.go
Comment (MSevey) [inline da/da.go:1]: can you revert the changes in this file? I'd like to see what happens with the tests with no changes to the current code.

Comment (arhamj) [reply da/da.go:1]: Reverted the changes in the da file. I have my earlier fixes to address the failing ones as a patch which is pretty clean. Let me know once you have enough data points. I can reapply the patch.

File: block/manager_test.go [hunk 3]
```diff
@@ -227,6 +227,48 @@ func TestSubmitBlocksToDA(t *testing.T) {
...
+				numBlocks, numTxs := 3, 5
+				blocks := make([]*types.Block, numBlocks)
+				blocks[0] = types.GetRandomBlock(uint64(1), numTxs)
+				blocks[1] = types.GetRandomBlock(uint64(2), numTxs)
+				blocks[2] = types.GetRandomBlock(uint64(3), numTxs)
+				invalidateBlockHeader(blocks[2])
+				return blocks
+			}(),
+			isErrExpected:               true,
+			expectedPendingBlocksLength: 1,
```
Comment (MSevey) [inline block/manager_test.go:242]:
```suggestion
			name: "A and B are fair blocks, but C has a marshalling error which results in no blocks being submitted.",
			blocks: func() []*types.Block {
				numBlocks, numTxs := 3, 5
				blocks := make([]*types.Block, numBlocks)
				blocks[0] = types.GetRandomBlock(uint64(1), numTxs)
				blocks[1] = types.GetRandomBlock(uint64(2), numTxs)
				blocks[2] = types.GetRandomBlock(uint64(3), numTxs)
				invalidateBlockHeader(blocks[2])
				return blocks
			}(),
			isErrExpected:               true,
			expectedPendingBlocksLength: 3,
```

**Review by MSevey — CHANGES_REQUESTED** [review]
Alright this is great. let's update the test case to have it pass and then I'll create some follow up issues to improve how these test cases are handled. At least now the full functionality is documented in these tests.

**Comment by Manav-Aggarwal** [comment]
Hey @arhamj, thanks for submitting this! It seems like the first test you added is failing in the unit test, could you make sure it's passing?

**Comment by arhamj** [comment]
> Hey @arhamj, thanks for submitting this! It seems like the first test you added is failing in the unit test, could you make sure it's passing?

@Manav-Aggarwal I had added a fix to handle the failing case, but @MSevey asked me to revert it to see what the failures were.

**Comment by MSevey** [comment]
> > Hey @arhamj, thanks for submitting this! It seems like the first test you added is failing in the unit test, could you make sure it's passing?
> 
> @Manav-Aggarwal I had added a fix to handle the failing case, but @MSevey asked me to revert it to see what the failures were.

@arhamj yes and then my last review was to update the test so that it passes with the current code. 
I captured the feature request for changing on to handle that test case here in #1590 since It will require some more discussion and I didn't want to block this PR any longer as it address the original issue request of adding the test cases.

**Comment by arhamj** [comment]
Sorry, I missed it completely that @MSevey 😄. Updating it.

File: block/manager_test.go (hunk 3 above)
Comment (MSevey) [reply block/manager_test.go:254]: still needs addressing plus it looks like their is a conflict with main now that needs to be addressed @arhamj

**Comment by arhamj** [comment]
Hey, sorry for the delay. @MSevey I think the marshalling error tests no longer make sense as we have gotten rid of the in-memory queue for the blocks. The store.SaveBlock() to add data to the DB takes care of MarshalErrors when writing and if there were a read error, store.GetBlock() would catch it. I've updated the PR. Let me know what you think.

File: block/manager_test.go [hunk 4]
```diff
@@ -282,6 +296,18 @@ func getTempKVStore(t *testing.T) ds.TxnDatastore {
 	return kvStore
 }
 
+// invalidateBlockHeader results in a block header that produces a marshalling error
+func invalidateBlockHeader(block *types.Block) {
```
Comment (tzdybal) [inline block/manager_test.go:300]: As marshaling errors are not tested now, let's remove this function.

**Review by tzdybal — CHANGES_REQUESTED** [review]
Please remove the unused function.

**Comment by MSevey** [comment]
> Hey, sorry for the delay. @MSevey I think the marshalling error tests no longer make sense as we have gotten rid of the in-memory queue for the blocks. The store.SaveBlock() to add data to the DB takes care of MarshalErrors when writing and if there were a read error, store.GetBlock() would catch it. I've updated the PR. Let me know what you think.

we should add this test back in. the reason is that we want to test functions in isolation. 
If there is an error path, we should test it, since external code is constantly changing. 
Not testing a code path based on assumptions of external code, leaves the code vulnerable to failing in the future if the external code changes again.

File: block/manager_test.go (hunk 4 above)
Comment (MSevey) [reply block/manager_test.go:300]: marshaling test case should just be added back in.

Comment (arhamj) [reply block/manager_test.go:367]: Added new tests compatible with the latest. I have also added a mock for the K-V Store using mockery.

One point to note is, before the recent updates, for the test condition: `Test_submitBlocksToDA_BlockMarshalErrorCase2`,
i.e. if A, B, and C are 3 pending blocks not yet submitted to the DA and block C has a marshal error, A and B get submitted. Now however none of them is submitted. (Rare event)

I'm using a more recent version of mockery do let me know if that's an issue.

File: block/manager_test.go [hunk 5]
```diff
@@ -282,6 +297,84 @@ func getTempKVStore(t *testing.T) ds.TxnDatastore {
 	return kvStore
 }
 
+// Test_submitBlocksToDA_BlockMarshalErrorCase1: A itself has a marshalling error. So A, B and C never get submitted.
+func Test_submitBlocksToDA_BlockMarshalErrorCase1(t *testing.T) {
```
Comment (MSevey) [inline block/manager_test.go:301]: @arhamj I think there is some confusion. I just wanted to add the marshal error cases back into `TestSubmitBlocksToDA`. 

We shouldn't need to add any new mocking or these separate tests. 

These tests you've added are testing that there is in fact an edge case in marshalling that gets by the reading from disk. That isn't want I was implying needs to be protected against. Between reading from disk and trying to marhsal  in submit blocks in submit blocks, the data struct is handled. This handling changes over time and could potentially introduce bugs or issues in the future. That is why we want to test the marshal edge case in submit blocks. 

We can remove these two tests and just put them back into `TestSubmitBlocksToDA` as test cases. 

You can also remove the added mock code as that shouldn't be needed.

File: Makefile [hunk 6]
```diff
@@ -84,6 +84,7 @@ mock-gen:
 	mockery --output test/mocks --srcpkg github.com/cometbft/cometbft/rpc/client --name Client
 	mockery --output test/mocks --srcpkg github.com/cometbft/cometbft/abci/types --name Application
 	mockery --output test/mocks --srcpkg github.com/rollkit/go-da --name DA
+	mockery --output test/mocks --srcpkg github.com/rollkit/rollkit/store --name Store
```
Comment (MSevey) [inline Makefile:87]: can be removed.

File: test/mocks/Store.go
Comment (MSevey) [inline test/mocks/Store.go:1]: Can be removed

**Review by MSevey — CHANGES_REQUESTED** [review]
Please revert back to just the new test cases to `TestSubmitBlocksToDA`. We don't need the new mocks or the separate tests. Only the new test cases.

File: block/manager_test.go (hunk 5 above)
Comment (tzdybal) [reply block/manager_test.go:301]: OK. This is a tough one.
We can't use normal store because it can't handle objects that can't be serialized / deserialized.
There is no place in the code where we can "invalidate" header retrieved from normal store.

I see two options here:
 * use mock store as proposed
 * refactor `submitBlocksToDA` to receive list of blocks instead of reading it from store; caller should read blocks from store and pass to `submitBlocksToDA` (or in case of test, we can just provide test-specific blocks)

Comment (MSevey) [reply block/manager_test.go:301]: OK, let's leave as is then.

**Review by MSevey — APPROVED** [review]

**Review by Manav-Aggarwal — APPROVED** [review]
